zero/src/provider.rs

```diff
@@ -7,9 +8,24 @@ use alloy::{providers::Provider, transports::Transport};
 use anyhow::Context;
 use tokio::sync::{Mutex, Semaphore, SemaphorePermit};
 
+use crate::rpc::RpcType;
+
 const CACHE_SIZE: usize = 1024;
 const MAX_NUMBER_OF_PARALLEL_REQUESTS: usize = 128;
 
+#[cfg(test)]
+use mockall::automock;
+
+#[cfg_attr(test, automock)]
+pub trait ZeroBlockProvider {
+    fn get_block_by_id(
+        &self,
+        block_id: BlockId,
+    ) -> impl Future<Output = anyhow::Result<Option<Block>>> + Send;
+
+    fn latest_block_number(&self) -> impl Future<Output = anyhow::Result<u64>> + Send;
+}
```

Comment:
I understand that this intermediate trait makes it easier to write tests, but the approach that feels right (tm) is to stick to `trait Provider` for composability.

How do you feel about sketching such a change?

Reply:
I've done this in https://github.com/0xPolygonZero/zk_evm/pull/745

Reply:
@0xaatif So you suggest we mock alloy `Provider` on the JSON response from the node level? It is very complex trait, and it is quite hard to mock it properly. For the tests, I don't see advantage in doing that.

Reply:
Related https://github.com/0xPolygonZero/zk_evm/pull/745#issuecomment-2431239933